plugins/discourse-presence/assets/javascripts/discourse/lib/presence-manager.js.es6

```diff
@@ -0,0 +1,205 @@
+// is tied closely with the server side implementation. Decreasing the duration
+// to increase the accuracy will come at the expense of having to more network
+// calls to publish the client's state.
+//
+// Logic walk through of our heuristic implementation:
+// - When client A is typing, a message is published every KEEP_ALIVE_DURATION_SECONDS.
+// - Client B receives the message and stores each user in an array and marks
+//   the user with a client-side timestamp of when the user was seen.
+// - If client A continues to type, client B will continue to receive messages to
+//   update the client-side timestamp of when client A was last seen.
+// - If client A disconnects or becomes inactive, the state of client A will be
+//   cleaned up on client B by a scheduler that runs every TIMER_INTERVAL_MILLISECONDS
+const KEEP_ALIVE_DURATION_SECONDS = 10;
+const BUFFER_DURATION_SECONDS = KEEP_ALIVE_DURATION_SECONDS + 2;
+
+const MESSAGE_BUS_LAST_ID = 0;
+const TIMER_INTERVAL_MILLISECONDS = 2000;
+
+export const REPLYING = "replying";
+export const EDITING = "editing";
+export const CLOSED = "closed";
+
+const PresenceManager = EmberObject.extend({
+  users: null,
+  editingUsers: null,
+  subscribed: null,
+  topic: null,
+  currentUser: null,
+  messageBus: null,
+  siteSettings: null,
+
+  init() {
+    this._super(...arguments);
+
+    this.setProperties({
+      users: [],
+      editingUsers: [],
+      subscribed: false
+    });
+  },
+
+  subscribe() {
+    if (this.subscribed) return;
+
+    this.messageBus.subscribe(
+      this.channel,
+      message => {
+        const { user, state } = message;
+        if (this.get("currentUser.id") === user.id) return;
+
+        switch (state) {
+          case REPLYING:
+            this._appendUser(this.users, user);
+            break;
+          case EDITING:
+            this._appendUser(this.editingUsers, user, {
+              post_id: parseInt(message.post_id, 10)
+            });
+            break;
+          case CLOSED:
+            this._removeUser(user);
+            break;
+        }
+      },
+      MESSAGE_BUS_LAST_ID
+    );
+
+    this.set("subscribed", true);
+  },
+
+  unsubscribe() {
+    this.messageBus.unsubscribe(this.channel);
+    this._stopTimer();
+    this.set("subscribed", false);
+  },
+
+  @discourseComputed("topic.id")
+  channel(topicId) {
+    return `/presence/${topicId}`;
+  },
+
+  throttlePublish(state, whisper, postId) {
+    return throttle(
+      this,
+      this.publish,
+      state,
+      whisper,
+      postId,
+      KEEP_ALIVE_DURATION_SECONDS * 1000
+    );
+  },
+
+  publish(state, whisper, postId) {
+    const data = {
+      state,
+      topic_id: this.get("topic.id")
+    };
+
+    if (whisper) {
+      data.is_whisper = 1;
+    }
+
+    if (postId) {
+      data.post_id = postId;
+    }
+
+    return ajax("/presence/publish", {
+      type: "POST",
+      data
+    });
+  },
+
+  _debounceSetProperties(properties) {
```

Comment:
where is this used?

Reply:
Oops good catch :+1: I thought I removed this.